src/tests/dotnet-trace/CollectCommandFunctionalTests.cs

```diff
@@ -0,0 +1,277 @@
+                    "Microsoft-Windows-DotNETRuntime 0x00000014C14FCCBD Informational(4) --profile")
+            };
+
+            yield return new object[]
+            {
+                new CollectArgs(output: CreateTraceFile("profileAndClrEventsIgnored"), profile: "cpu-sampling", clrevents: "gc"),
+                ExpectProvidersWithMessages(
+                    new[]
+                    {
+                        "The argument --clrevents gc will be ignored because the CLR provider was configured via either --profile or --providers command."
+                    },
+                    "Microsoft-DotNETCore-SampleProfiler 0x0000F00000000000 Informational(4) --profile",
+                    "Microsoft-Windows-DotNETRuntime 0x00000014C14FCCBD Informational(4) --profile")
+            };
+
+            yield return new object[]
+            {
+                new CollectArgs(output: CreateTraceFile("profileOverriddenRuntime"), profile: "cpu-sampling", providers: "Microsoft-Windows-DotNETRuntime:0x1:4"),
+                ExpectProviders(
+                    "Microsoft-Windows-DotNETRuntime 0x0000000000000001 Informational(4) --providers",
+                    "Microsoft-DotNETCore-SampleProfiler 0x0000F00000000000 Informational(4) --profile")
+            };
+
+            yield return new object[]
+            {
+                new CollectArgs(output: CreateTraceFile("providersClrEventsIgnored"), providers: "Microsoft-Windows-DotNETRuntime:0x1:4", clrevents: "gc"),
+                ExpectProvidersWithMessages(
+                    new[]
+                    {
+                        "The argument --clrevents gc will be ignored because the CLR provider was configured via either --profile or --providers command."
+                    },
+                    "Microsoft-Windows-DotNETRuntime 0x0000000000000001 Informational(4) --providers")
+            };
+
+            yield return new object[]
+            {
+                new CollectArgs(output: CreateTraceFile("clrEventsOnly"), clrevents: "gc+jit"),
+                ExpectProviders(
+                    "Microsoft-Windows-DotNETRuntime 0x0000000000000011 Informational(4) --clrevents")
+            };
+
+            yield return new object[]
+            {
+                new CollectArgs(output: CreateTraceFile("clrEventsVerbose"), clrevents: "gc+jit", clreventlevel: "5"),
+                ExpectProviders(
+                    "Microsoft-Windows-DotNETRuntime 0x0000000000000011 Verbose(5) --clrevents")
+            };
+        }
+
+        private static FileInfo CreateTraceFile(string prefix)
+        {
+            string fullPath = Path.Combine(TestTraceDir, $"{prefix}_{Guid.NewGuid():N}.nettrace");
+            if (File.Exists(fullPath))
+            {
+                File.Delete(fullPath);
+            }
+            return new FileInfo(fullPath);
+        }
+
+        private static readonly string TestTraceDir = CreateTestTraceDir();
```

Comment:
Ideally we'd avoid writing any files in these tests. We can do it if we have to but it comes with potential issues like failing to clean up the files after the test is complete, dealing with IO failures, and some extra perf overhead.

I'd suggest extending the test seam so that the test can intercept the call to new FileStream and return a MemoryStream instead. You probably also need to intercept checking the length of the file as it is being written.

Reply:
I initially figured that the files wouldn't be large + allowed inspection upon test failure in case something went awry with copying the stream. I wasn't sure if we needed our own logic to clean up artifacts, or if our CI testing infra handled them in some way.

I added the seam to allow for dotnet-trace collect to leverage a FileStream and for the unit tests to leverage a MemoryStream.